File: test/pytest/test_publisher.py
#!/usr/bin/env python3

"""
Test script for the ROS 2 minimal publisher node.

This script contains the unit test for verifying a minimal ROS 2 publisher.
It tests the node creation, message counter increment, and message content formatting.

-------
Subscription Topics:
    None

-------
Publishing Topics:
    /py_example_topic (std_msgs/String): Example messages with incrementing counter.

:author: Alessandro Tiozzo
:date: June 23, 2026
"""

import pytest
import rclpy
from std_msgs.msg import String
from ros2_fundamentals_examples.py_minimal_publisher import MinimalPyPublisher

def test_publisher_creation():
    """
    Test if the publisher node is created correctly.
    
    This test verifies:
    1. The node name is set correctly.
    2. The publisher object exists.
    3. The topic name is correct.
    
    :raises: AssertionError if any of the checks fail.
    """

    rclpy.init()
    try:
        # Create an instance of the publisher node
        node = MinimalPyPublisher()

        # Test 1: Verify the node name
        assert node.get_name() == 'py_minimal_publisher', "Node name is incorrect."

        # Test 2: Verify the publisher exists and has the correct topic name
        assert hasattr(node, 'publisher_1'), "Publisher object does not exist."
        assert node.publisher_1.topic_name == '/py_example_topic', "Publisher topic name is incorrect."
    
    finally:
        # Clean up ROS 2 communication
        rclpy.shutdown()


def test_message_counter_increment():
    """
    Test if the message counter increments correctly.

    This test verifies that the counter (node.i) increases by 1 after each
    timer callback execution.

    :raises: AssertionError if the counter does not increment as expected.
    """

    rclpy.init()
    try:
        # Create an instance of the publisher node
        node = MinimalPyPublisher()

        # Store the initial counter value
        initial_counter = node.i

        # Simulate a timer callback to publish a message
        node.timer_callback()

        # Test: Verify the counter has incremented by 1
        assert node.i == initial_counter + 1, "Message counter did not increment correctly."
    
    finally:
        # Clean up ROS 2 communication
        rclpy.shutdown()


def test_message_content():
    """
    Test if the published message content is formatted correctly.

    This test verifies that the message data follows the expected format
    "Hello World: <counter_value>".

    :raises: AssertionError if the message content is not as expected.
    """

    rclpy.init()
    try:
        # Create an instance of the publisher node
        node = MinimalPyPublisher()

        # Set counter to a known value for testing
        node.i = 5
        msg = String()

        # Retrieve the last published message
        msg.data = f'Hello World: {node.i}'

        # Test: Verify the message content format
        assert msg.data == 'Hello World: 5', "Published message content is incorrect."
    
    finally:
        # Clean up ROS 2 communication
        rclpy.shutdown()

if __name__ == '__main__':
    # Run the tests if the script is executed directly
    pytest.main(['-v'])

File: test/gtest/test_publisher.cpp
/**
 * @file test_publisher.cpp
 * @brief Unit test for the ROS 2 minimal publisher node.
 * 
 * This file contains test cases to verify the functionality of our minimal publisher node.
 * We test two main things:
 * 1. The publisher node can be created successfully with the right name and topic.
 * 2. The publisher node can publish messages with "Hello World: <counter_value>".
 * 
 * Testing framework:
 *  Google Test (gtest) is used for writing and running the unit tests in C++.
 * 
 * Tests:
 *  TestNodeCreation: Verifies node name and publisher setup.
 *  TestMessageContent: Verifies published message format.
 * 
 * @author Alessandro Tiozzo
 * @date 2026-06-24
*/

#include <gtest/gtest.h>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class CppMinimalPublisher;

// Exclude the main function in the publisher node during testing
#define TESTING_EXCLUDE_MAIN 
#include "../../src/cpp_minimal_publisher.cpp"


class TestMinimalPublisher : public ::testing::Test {
protected:
    void SetUp() override {
        // Initialize ROS 2
        rclcpp::init(0, nullptr);

        node = std::make_shared<CppMinimalPublisher>();
    }

    void TearDown() override {
        // Shutdown ROS 2
        rclcpp::shutdown();
    }

    std::shared_ptr<CppMinimalPublisher> node;
};

TEST_F(TestMinimalPublisher, TestNodeCreation) {
    // Verify that the node was created successfully
    EXPECT_EQ(std::string(node->get_name()), std::string("cpp_minimal_publisher"));

    // Verify that the publisher is set up correctly and is unique
    auto pub_endpoints = node->get_publishers_info_by_topic("/cpp_example_topic");
    EXPECT_EQ(pub_endpoints.size(), 1);
}

TEST_F(TestMinimalPublisher, TestMessageContent) {
    // Create a message and publish it
    std::shared_ptr<std_msgs::msg::String> received_message;
    
    auto subscription = node->create_subscription<std_msgs::msg::String>(
        "/cpp_example_topic",
        10,
        [&received_message](std_msgs::msg::String::SharedPtr msg) {
            received_message = std::make_shared<std_msgs::msg::String>(*msg);
        }
    );

    node->timer_callback(); // Trigger the timer callback to publish a message
    rclcpp::spin_some(node); // Process the callback to receive the message

    // Verify that the published message has the expected content
    EXPECT_EQ(received_message->data.substr(0,12), "Hello, world");
}

int main(int argc, char** argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}